app/upgrades/v17/convert.go

```diff
@@ -0,0 +1,159 @@
+// Copyright Tharsis Labs Ltd.(Evmos)
+// SPDX-License-Identifier:ENCL-1.0(https://github.com/evmos/evmos/blob/main/LICENSE)
+
+package v17
+
+import (
+	"fmt"
+	"math/big"
+
+	errorsmod "cosmossdk.io/errors"
+	"github.com/cometbft/cometbft/libs/log"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	authkeeper "github.com/cosmos/cosmos-sdk/x/auth/keeper"
+	authtypes "github.com/cosmos/cosmos-sdk/x/auth/types"
+	bankkeeper "github.com/cosmos/cosmos-sdk/x/bank/keeper"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/evmos/evmos/v16/contracts"
+	"github.com/evmos/evmos/v16/precompiles/werc20/testdata"
+	erc20keeper "github.com/evmos/evmos/v16/x/erc20/keeper"
+	erc20types "github.com/evmos/evmos/v16/x/erc20/types"
+	evmtypes "github.com/evmos/evmos/v16/x/evm/types"
+)
+
+// ConvertERC20Coins converts Native IBC coins from their ERC20 representation
+// to the native representation. This also includes the withdrawal of WEVMOS tokens
+// to EVMOS native tokens.
+func ConvertERC20Coins(
+	ctx sdk.Context,
+	logger log.Logger,
+	accountKeeper authkeeper.AccountKeeper,
+	bankKeeper bankkeeper.Keeper,
+	erc20Keeper erc20keeper.Keeper,
+	wrappedAddr common.Address,
+	nativeTokenPairs []erc20types.TokenPair,
+) error {
+	// iterate over all the accounts and convert the tokens to native coins
+	accountKeeper.IterateAccounts(ctx, func(account authtypes.AccountI) (stop bool) {
+		cosmosAddress := account.GetAddress()
+		ethAddress := common.BytesToAddress(cosmosAddress.Bytes())
+		ethHexAddr := ethAddress.String()
+
+		balance, res, err := WithdrawWEVMOS(ctx, ethAddress, wrappedAddr, erc20Keeper)
+
+		var bs string // NOTE: this is necessary so that there is no panic if balance is nil when logging
+		if balance != nil {
+			bs = balance.String()
+		}
+
+		if err != nil {
+			logger.Error(
+				"failed to withdraw WEVMOS",
+				"account", ethHexAddr,
+				"balance", bs,
+				"error", err.Error(),
+			)
+		} else if res != nil && res.VmError != "" {
+			logger.Error(
+				"withdraw WEVMOS reverted",
+				"account", ethHexAddr,
+				"balance", bs,
+				"vm-error", res.VmError,
+			)
+		}
+
+		for _, tokenPair := range nativeTokenPairs {
+			contract := tokenPair.GetERC20Contract()
+			if err := ConvertERC20Token(ctx, ethAddress, contract, cosmosAddress, erc20Keeper); err != nil {
```

Comment:
This function is not only doing conversion. It also check if the current account has the balance. A bit misleading the name